### PR TITLE
Make updates to serp events table

### PR DESCRIPTION
**Pull request by dzeber**
Update the serp_events table design to fix some issues and make it easier to use. [This query](https://console.cloud.google.com/bigquery?sq=687003905036:011b738339b7402b9917757949d881c6) gives a sample of how the result of querying the new view would look.

Summary of changes:
- 1 row per impression ID with ad impression & engagement info grouped into arrays (previous version has multiple rows per impression ID)
- Count all engagement actions rather than specific ones. New engagement actions added to the events later on will automatically get included in the table.
- Mapping logic moved to view layer
- Change to the logic in counting visible ads following updated discussions with Eng.
    - previously, we used `num_ads_showing` to count visible ads, depending on the ad blocker inference logic.
    - we now use reported `ads_visible` as the count of visible ads, and `num_ads_showing` column is removed
    - `_reported` suffix dropped from ad impression counts
    - change naming of `_hidden` to `_blocked`
- Change naming in table columns from 'ad impressions' to 'component/element impressions' to reflect broader usage of `ad_impression` event
- Add summary columns for impression-level ad impressions & engagements

Jira ticket: [RS-993](https://mozilla-hub.atlassian.net/browse/RS-993)

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

[RS-993]: https://mozilla-hub.atlassian.net/browse/RS-993?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-3269)

### CODE REVIEW

**Comment by jaredsnyder** [comment]
Could you provide more detail on what "replacing" `num_ads_showing` with `ads_visible`?  Are we essentially choosing to not consider a special case where some but not all ads are hidden?

File: sql_generators/serp_events/templates/desktop_query.sql [hunk 1]
```diff
@@ -193,64 +188,25 @@ SELECT
   sample_id,
   experiments,
   is_shopping_page,
+  is_private,
```
Comment (jaredsnyder) [inline sql_generators/serp_events/templates/desktop_query.sql:191]: Does this mean we have info about PBM now?

**Review by jaredsnyder — COMMENTED** [review]
I don't feel qualified to "Approve" but as a consumer it all looks good.  My one concern is that the DTMO docs, especially the example queries, get updated once this is merged but I don't know what that process looks like.

File: sql_generators/serp_events/templates/desktop_query.sql (hunk 1 above)
Comment (dzeber) [reply sql_generators/serp_events/templates/desktop_query.sql:191]: yes, this was [added](https://bugzilla.mozilla.org/show_bug.cgi?id=1851495) in 120.

Comment (jaredsnyder) [reply sql_generators/serp_events/templates/desktop_query.sql:191]: Nice!

**Comment by dzeber** [comment]
> Could you provide more detail on what "replacing" `num_ads_showing` with `ads_visible`? Are we essentially choosing to not consider a special case where some but not all ads are hidden?

This is per the latest guidance from Eng. There are a couple reasons why we would could get some ads visible and some hidden, which we can't currently distinguish between, and neither of which we expect to be that common:
1. it could be an ad blocker selectively allowing some ads through
2. it could be an issue with the detection counting an ad link as hidden either erroneously or when it's actually associated with a different, visible ad on the page.

In case (1), `ads_visible` is the correct count. In case (2), when it's associated with a visible ad, then `ads_visible` would be the correct count and adding `ads_hidden` would actually double-count.
So we figure that relying on `ads_visible` at face value is the best approach, absent more specific info on which situation is occurring when.

> My one concern is that the DTMO docs, especially the example queries, get updated once this is merged but I don't know what that process looks like.

There's no automated process to enforce this, it's more of a case where I have a related ticket for updating docs that I plan to address after :).

File: sql_generators/serp_events/templates/desktop_query.sql (hunk 1 above)
Comment (pissac17) [reply sql_generators/serp_events/templates/desktop_query.sql:191]: @dzeber Are there any metrics that we should not be tracking in PBM, or are we tracking everything just like in non-PBM mode?

**Review by pissac17 — APPROVED** [review]

File: sql_generators/serp_events/templates/desktop_query.sql (hunk 1 above)
Comment (dzeber) [reply sql_generators/serp_events/templates/desktop_query.sql:191]: Not sure, I haven't followed the discussion on this update or dug into the data. I saw this flag was added to the event so I wanted to capture it in this table.

File: sql_generators/serp_events/templates/desktop_query.sql [hunk 2]
```diff
@@ -145,39 +152,27 @@ ad_impression_counts AS (
...
+      STRUCT(
+        component,
+        ads_loaded AS num_elements_loaded,
+        ads_visible AS num_elements_visible,
+        -- elements explicitly hidden using JS/CSS, as by an ad blocker
+        ads_hidden AS num_elements_blocked,
+        -- elements that are not visible but not explicitly blocked, eg. below the fold
+        ads_loaded - ads_visible - ads_hidden AS num_elements_notshowing
+      )
+    ) AS component_impressions
```
[the remark this answers is not in the document]
Comment (dzeber) [reply sql_generators/serp_events/templates/desktop_query.sql:171]: I'm fine with making this change. My only concern would be confusion with top-level "SERP impressions" and the impression_id, but maybe this wouldn't be a major issue.

File: sql_generators/serp_events/templates/view.sql [hunk 3]
```diff
@@ -1,8 +1,85 @@
---- User-facing view. Generated via sql_generators.active_users.
+CREATE TEMP FUNCTION is_ad_component(component STRING) AS (
```
[the remark this answers is not in the document]
Comment (dzeber) [reply sql_generators/serp_events/templates/view.sql:1]: Sounds good, I'll update that

[the remark this answers is not in the document]
Comment (dzeber) [reply sql_generators/serp_events/templates/view.sql]: @alekhyamoz I moved both of these functions to mozfun

**Review by alekhyamoz — APPROVED** [review]
LGTM

**Comment by dzeber** [comment]
These changes have been split off across #5417 and #5427

**Pull request closed without merging**